source/connection-troubleshooting.txt

```diff
@@ -0,0 +1,308 @@
+--------------------------
+
+Verify that you've accessed the MongoDB deployment with the correct user. The
+term "message" in the error can be a command sent by the driver.
+If you are using a user that doesn't have permissions to send the command, the
+driver could generate this error.
+
+Also ensure that the user has the appropriate permissions for the message you
+are sending. MongoDB uses Role-Based Access Control (RBAC) to control access
+to a MongoDB deployment. For more information about how to configure RBAC in MongoDB,
+see :manual:`Default MongoDB Port </core/authorization/>`.
+
+Configure Your Firewall
+-----------------------
+
+The firewall needs to have an open port for communicating with the MongoDB
+instance. For more information about configuring the firewall, see
+:ref:`Configure Your Firewall <csharp-troubleshooting-connection-firewall>` in
+the Connection Error section.
+
+.. _csharp-troubleshooting-connection-number-connections:
+
+Check the Number of Connections
+-------------------------------
+
+Each ``MongoClient`` instance supports a maximum number of concurrent open
+connections in its connection pool. You can configure the parameter ``MaxConnectionPoolSize``
+which defines this limit. The default value is ``100``. If there are already a
+number of open connections equal to ``MaxConnectionPoolSize``, the server waits until
+a connection becomes available. If this wait time exceeds the ``MaxConnectionIdleTime``
+value, the driver responds with an error.
+
+For more information about how connection pooling works, see
+:ref:`How Does Connection Pooling Work in the {+driver-short+}? <csharp-faq-connection-pool>`
+in the FAQ.
+
+Too Many Open Connections
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The driver creates the following error message when it attempts to open a
+connection, but it's reached the maximum number of connections:
+
+.. code-block:: none
+   :copyable: false
+
+   connection refused because too many open connections
+
+The following section describes a method that may help resolve the issue.
+
+Check the Number of Connections
+-------------------------------
+
+If you need to create more open connections, increase ``MaxConnectionPoolSize``. For more
+information about checking the number of connections, see
+:ref:`Check the Number of Connections <csharp-troubleshooting-connection-number-connections>`
+in the Error Sending Message section.
+
+Timeout Error
+~~~~~~~~~~~~~
+
+When the network is not able to deliver a request from the driver to the server
+quickly enough, it can time out. When this happens, you might receive an error message
+similar to the following message:
+
+.. code-block:: none
+   :copyable: false
+
+   timed out while checking out a connection from connection pool: context canceled
+
+If you receive this error, try the following action to resolve the
+issue.
+
+Set connectTimeoutMS
+--------------------
+
+The driver may hang when it's unable to establish a connection because it
```

Comment:
S: the "it" at the end of this line is a little bit unclear. Clarify if it refers to the driver or the connection